### PR TITLE
Minor: Fix incorrect indices for hashing struct

### PR DESCRIPTION
**Pull request by jayzhan211**
## Which issue does this PR close?

Bug in https://github.com/apache/arrow-datafusion/pull/8552
closes https://github.com/apache/arrow-datafusion/pull/8769

## Are these changes tested?

Failed test is added

### CODE REVIEW

File: datafusion/common/src/hash_utils.rs [hunk 1]
```diff
@@ -214,22 +214,19 @@ fn hash_struct_array(
     hashes_buffer: &mut [u64],
 ) -> Result<()> {
     let nulls = array.nulls();
-    let num_columns = array.num_columns();
+    let row_len = array.len();
 
-    // Skip null columns
```
Comment (jayzhan211) [inline datafusion/common/src/hash_utils.rs:219]: I forgot to correct the comment last time. Indices are all about row, not columns

File: datafusion/common/src/hash_utils.rs [hunk 2]
```diff
@@ -214,22 +214,19 @@ fn hash_struct_array(
...
-    let valid_indices: Vec<usize> = if let Some(nulls) = nulls {
+    let valid_row_indices: Vec<usize> = if let Some(nulls) = nulls {
         nulls.valid_indices().collect()
     } else {
-        (0..num_columns).collect()
```
Comment (jayzhan211) [inline datafusion/common/src/hash_utils.rs:223]: Main fix

File: datafusion/sqllogictest/test_files/dictionary.slt [hunk 3]
```diff
@@ -148,7 +148,7 @@ select count(*) from m1 where tag_id = '1000' and time < '2024-01-03T14:46:35+01
 ----
 10
 
-query RRR
+query RRR rowsort
```
Comment (jayzhan211) [inline datafusion/sqllogictest/test_files/dictionary.slt:151]: rowsort is added because I got
```
700 1000 850
100 600 350
```

Comment (alamb) [reply datafusion/sqllogictest/test_files/dictionary.slt:151]: Yeah --  also added the same fix in https://github.com/apache/arrow-datafusion/pull/8769 and I agree this looks good

**Review by alamb — APPROVED** [review]
Thank you @jayzhan211

**Pull request merged**